keadm/cmd/keadm/app/cmd/util/helminstaller.go

```diff
@@ -0,0 +1,501 @@
+	// DefaultProfileFilename is the name of the default profile yaml file.
+	DefaultProfileFilename = "version.yaml"
+	DefaultHelmValuesPath  = "values.yaml"
+
+	DefaultHelmTimeout = time.Duration(60 * time.Second)
+
+	DefaultHelmInstall  = true
+	DefaultHelmWait     = true
+	DefaultHelmCreateNs = true
+)
+
+var (
+	ErrListProfiles = errors.New("can not list profiles")
+)
+
+// KubeCloudHelmInstTool embeds Common struct
+// It implements ToolsInstaller interface
+type KubeCloudHelmInstTool struct {
+	Common
+	AdvertiseAddress string
+	Manifests        string
+	Namespace        string
+	CloudcoreImage   string
+	CloudcoreTag     string
+	IptablesMgrImage string
+	IptablesMgrTag   string
+	Sets             []string
+	Profile          string
+	ProfileKey       string
+	ExternalHelmRoot string
+	Force            bool
+	SkipCRDs         bool
+	DryRun           bool
+	Action           string
+	existsProfile    bool
+}
+
+// InstallTools downloads KubeEdge for the specified version
+// and makes the required configuration changes and initiates cloudcore.
+func (cu *KubeCloudHelmInstTool) InstallTools() error {
+	cu.SetOSInterface(GetOSInterface())
+	cu.SetKubeEdgeVersion(cu.ToolVersion)
+
+	baseHelmRoot := DefaultBaseHelmDir
+	if cu.ExternalHelmRoot != "" {
+		baseHelmRoot = cu.ExternalHelmRoot
+	}
+
+	switch cu.Action {
+	case types.HelmInstallAction:
+		if err := cu.RunHelmInstall(baseHelmRoot); err != nil {
+			return err
+		}
+	case types.HelmManifestAction:
+		if err := cu.RunHelmManifest(baseHelmRoot); err != nil {
+			return err
+		}
+	default:
+		fmt.Println("Not support this action")
+	}
+
+	return nil
+}
+
+// RunHelmInstall renders the Charts with the given values, then installs the Charts to the cluster.
+func (cu *KubeCloudHelmInstTool) RunHelmInstall(baseHelmRoot string) error {
+	// --force would not care about whether the cloud components exist or not
+	// Also, if gives a external helm root, no need to check and verify. Because it is always not a cloudcore.
+	if !cu.Force && cu.ExternalHelmRoot == "" {
+		cloudCoreRunning, err := cu.IsKubeEdgeProcessRunning(KubeCloudBinaryName)
+		if err != nil {
+			return err
+		}
+		if cloudCoreRunning {
+			return fmt.Errorf("CloudCore is already running on this node, please run reset to clean up first")
+		}
+	}
+
+	err := cu.IsK8SComponentInstalled(cu.KubeConfig, cu.Master)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Kubernetes version verification passed, KubeEdge installation will start...")
+
+	// prepare to render
+	if err := cu.beforeRenderer(baseHelmRoot); err != nil {
+		return err
+	}
+
+	// build a renderer instance with the given values and flagvals
+	renderer, err := cu.buildRenderer(baseHelmRoot)
+	if err != nil {
+		return fmt.Errorf("cannot build renderer: %s", err.Error())
+	}
+
+	release, err := cu.runHelmInstall(renderer)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s started\n", strings.ToTitle(renderer.componentName))
+
+	if release == nil {
+		return nil
```

Comment:
return error?

Reply:
will fix these

Reply:
Have handled before the code.
```
	release, err := cu.runHelmInstall(renderer)
	if err != nil {
		return err
	}

	fmt.Printf("%s started\n", strings.ToTitle(renderer.componentName))

	if release == nil {
		return nil
	}
```

Reply:
These codes prevent the empty point and are only used for print.

Reply:
fixed